--- ../../pikeyd/config.h	2021-04-28 11:14:37.810338339 -0400
+++ config.h	2021-04-14 06:56:38.346567202 -0400
@@ -26,37 +26,26 @@
 #ifndef _CONFIG_H_
 #define _CONFIG_H_
 
-#include "iic.h"
-
 typedef struct{
   char name[32];
   int code;
 }key_names_s;
 
-typedef struct{
+typedef struct {
   int key;
-  int val;
-}keyinfo_s;
+  long unsigned int bits;
+  long unsigned int pressed;
+}key_special_s;
+
 
 int init_config(void);
 void test_config(void);
 int get_event_key(int gpio, int idx);
 int get_next_key(int gpio);
 int got_more_keys(int gpio);
-int got_more_xio_keys(int xio, int gpio);
 void restart_keys(void);
 int gpios_used(void);
 int gpio_pin(int n);
-int is_xio(int gpio);
-int get_curr_key(void);
-int get_curr_xio_no(void);
-void get_xio_parm(int xio, iodev_e *type, int *addr, int *regno);
-int get_next_xio_key(int xio, int gpio);
-void restart_xio_keys(int xio);
-void handle_iic_event(int xio, int value);
-void last_iic_key(keyinfo_s *kp);
-
-
 
 #endif
 
